pkg/ipcache/metadata.go

```diff
@@ -218,27 +236,31 @@ func (ipc *IPCache) InjectLabels(ctx context.Context, modifiedPrefixes []netip.P
 				break
 			}
 
-			// We can safely skip the ipcache upsert if the ID and source
-			// in the metadata cache match the ipcache exactly.
-			// Note that checking ID alone is insufficient, see GH-24502
-			if oldID.ID == newID.ID && prefixInfo.Source() == oldID.Source {
+			// We can safely skip the ipcache upsert if the entry matches with
+			// the entry in the metadata cache exactly.
+			// Note that checking ID alone is insufficient, see GH-24502.
+			if oldID.ID == newID.ID && prefixInfo.Source() == oldID.Source &&
+				oldTunnelIP.Equal(prefixInfo.TunnelPeer().IP()) &&
+				oldEncryptionKey == prefixInfo.EncryptKey().Uint8() {
 				goto releaseIdentity
 			}
```

Comment:
Point this out as a key diff from previous revision of the PR

Reply:
cc @gandro @joestringer, this was needed to resolve the encryption issue on upgrade on kind clusters (with tunneling), part 1/2.

Reply:
Good catch! Should we add a `prefixInfo.HasChanged(other)` method that performs this the check on auxiliary (i.e. metadata that does not change the ID)? We might add more non-identity defining metadata in the future (e.g. named ports) that will very likely need the same check here.

Reply:
Yes good idea. I'll add it to https://github.com/cilium/cilium/issues/21142.